File: Android/app/src/main/java/com/edu/primary/repository/UserRepository.java
package com.edu.primary.repository;

import android.content.Context;
import android.content.SharedPreferences;
import com.edu.primary.constants.AppConstants;
import com.edu.primary.database.AppDatabase;
import com.edu.primary.database.entity.UserEntity;
import com.edu.primary.utils.InputValidator;
import com.edu.primary.utils.Logger;
import com.edu.primary.utils.PasswordUtil;
import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;

public class UserRepository {

    private final AppDatabase database;
    private final SharedPreferences prefs;

    /**
     * 构造函数
     * 使用ApplicationContext防止内存泄漏
     */
    public UserRepository(Context context) {
        // 使用ApplicationContext防止内存泄漏
        Context appContext = context.getApplicationContext();
        database = AppDatabase.getInstance(appContext);
        prefs = appContext.getSharedPreferences(AppConstants.PREFS_USER, Context.MODE_PRIVATE);
    }

    /**
     * 注册用户
     * 添加输入验证和密码加密
     */
    public Single<Long> register(String username, String password, String nickname) {
        return Single.fromCallable(() -> {
            // 输入验证
            if (!InputValidator.isValidUsername(username)) {
                Logger.w("UserRepository", "Invalid username: " + username);
                throw new IllegalArgumentException("用户名格式不正确（3-20个字符，只能包含字母、数字、下划线）");
            }
            if (!InputValidator.isValidPassword(password)) {
                Logger.w("UserRepository", "Invalid password length");
                throw new IllegalArgumentException("密码长度至少6个字符");
            }
            
            UserEntity existingUser = database.userDao().getUserByUsername(username);
            if (existingUser != null) {
                Logger.w("UserRepository", "Username already exists: " + username);
                throw new Exception(AppConstants.ERROR_USERNAME_EXISTS);
            }
            
            // 加密密码
            String encryptedPassword = PasswordUtil.encrypt(password);
            UserEntity user = new UserEntity(username, encryptedPassword, nickname);
            long userId = database.userDao().insertUser(user);
            Logger.d("UserRepository", "User registered: " + username);
            return userId;
        }).subscribeOn(Schedulers.io());
    }

    /**
     * 用户登录
     * 添加输入验证和密码验证
     */
    public Single<UserEntity> login(String username, String password) {
        return Single.fromCallable(() -> {
            // 输入验证
            if (!InputValidator.isValidUsername(username)) {
                Logger.w("UserRepository", "Invalid username: " + username);
                throw new IllegalArgumentException("用户名格式不正确");
            }
            if (!InputValidator.isValidPassword(password)) {
                Logger.w("UserRepository", "Invalid password");
                throw new IllegalArgumentException("密码格式不正确");
            }
            
            UserEntity user = database.userDao().getUserByUsername(username);
            if (user == null) {
                Logger.w("UserRepository", "User not found: " + username);
                throw new Exception(AppConstants.ERROR_USER_NOT_FOUND);
            }
            
            // 验证密码（支持加密和明文密码，便于迁移）
            boolean passwordMatch = PasswordUtil.verify(password, user.password) 
                || user.password.equals(password); // 兼容旧数据
            
            if (!passwordMatch) {
                Logger.w("UserRepository", "Password mismatch for user: " + username);
                throw new Exception(AppConstants.ERROR_PASSWORD_WRONG);
            }
            
            // 如果密码是明文的，更新为加密密码
            if (user.password.equals(password)) {
                user.password = PasswordUtil.encrypt(password);
                database.userDao().updateUser(user);
                Logger.d("UserRepository", "Password encrypted for user: " + username);
            }
            
            saveCurrentUser(user.id, user.username);
            Logger.d("UserRepository", "User logged in: " + username);
            return user;
        }).subscribeOn(Schedulers.io());
    }

    public void logout() {
        prefs.edit().clear().apply();
    }

    public boolean isLoggedIn() {
        return prefs.getLong(AppConstants.KEY_USER_ID, -1) != -1;
    }

    public long getCurrentUserId() {
        return prefs.getLong(AppConstants.KEY_USER_ID, -1);
    }

    public String getCurrentUsername() {
        return prefs.getString(AppConstants.KEY_USERNAME, "");
    }

    private void saveCurrentUser(long userId, String username) {
        prefs.edit()
            .putLong(AppConstants.KEY_USER_ID, userId)
            .putString(AppConstants.KEY_USERNAME, username)
            .apply();
    }

    public Single<UserEntity> getCurrentUser() {
        return Single.fromCallable(() -> {
            long userId = getCurrentUserId();
            if (userId == -1) {
                return null;
            }
            return database.userDao().getUserById(userId);
        }).subscribeOn(Schedulers.io());
    }
}

File: Android-1/app/src/main/java/com/edu/primary/data/repository/UserRepository.kt
package com.edu.primary.data.repository

import android.content.Context
import com.edu.primary.data.local.AppDatabase
import com.edu.primary.data.local.entity.UserEntity
import com.edu.primary.util.AppConstants
import com.edu.primary.util.InputValidator
import com.edu.primary.util.Logger
import com.edu.primary.util.PasswordUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext

/**
 * 用户相关数据仓库：登录、注册、当前用户状态。
 *
 * 使用 Room [AppDatabase.userDao] 持久化用户，使用 SharedPreferences 保存当前登录用户 ID 与用户名。
 * 密码经 [PasswordUtil.encrypt] 存储；登录时支持明文兼容并自动升级为加密存储。
 *
 * @param context 用于获取 ApplicationContext 与 SharedPreferences，建议传入 Application
 */
class UserRepository(context: Context) {
    private val appContext = context.applicationContext
    private val db = AppDatabase.getInstance(appContext)
    private val prefs = appContext.getSharedPreferences(AppConstants.PREFS_USER, Context.MODE_PRIVATE)

    private val _currentUserId = MutableStateFlow(prefs.getLong(AppConstants.KEY_USER_ID, -1L))
    /** 当前登录用户 ID 的响应式流，未登录时为 -1L。 */
    val currentUserId: StateFlow<Long> = _currentUserId.asStateFlow()

    /** 是否已登录（即 SharedPreferences 中存有有效 user_id）。 */
    fun isLoggedIn(): Boolean = prefs.getLong(AppConstants.KEY_USER_ID, -1L) != -1L
    /** 当前用户 ID，未登录返回 -1L。 */
    fun getCurrentUserId(): Long = prefs.getLong(AppConstants.KEY_USER_ID, -1L)
    /** 当前用户名，未登录返回空字符串。 */
    fun getCurrentUsername(): String = prefs.getString(AppConstants.KEY_USERNAME, "").orEmpty()

    /** 清除登录状态（清空 SharedPreferences 并更新 [currentUserId]）。 */
    fun logout() {
        prefs.edit().clear().apply()
        _currentUserId.value = -1L
    }

    private fun saveCurrentUser(userId: Long, username: String) {
        prefs.edit()
            .putLong(AppConstants.KEY_USER_ID, userId)
            .putString(AppConstants.KEY_USERNAME, username)
            .apply()
        _currentUserId.value = userId
    }

    /**
     * 注册新用户。密码会经 MD5 加密后存入数据库。
     * @param username 用户名，须符合 [InputValidator.isValidUsername]
     * @param password 密码，须符合 [InputValidator.isValidPassword]
     * @param nickname 昵称，可与 username 相同
     * @return [Result.success] 为新用户 ID，[Result.failure] 为校验失败或用户名已存在等异常
     */
    suspend fun register(username: String, password: String, nickname: String): Result<Long> = withContext(Dispatchers.IO) {
        if (!InputValidator.isValidUsername(username))
            return@withContext Result.failure(IllegalArgumentException("用户名格式不正确（3-20个字符，字母数字下划线）"))
        if (!InputValidator.isValidPassword(password))
            return@withContext Result.failure(IllegalArgumentException("密码至少6位"))
        val existing = db.userDao().getByUsername(username)
        if (existing != null)
            return@withContext Result.failure(Exception(AppConstants.ERROR_USERNAME_EXISTS))
        val encrypted = PasswordUtil.encrypt(password)
        val user = UserEntity(username = username, password = encrypted, nickname = nickname)
        val id = db.userDao().insert(user)
        Logger.d("UserRepository", "Registered: $username")
        Result.success(id)
    }

    /**
     * 用户登录。校验用户名与密码，成功后写入当前用户到 SharedPreferences 并更新 [currentUserId]。
     * @param username 用户名
     * @param password 密码（支持与库中明文兼容，登录后会自动更新为加密存储）
     * @return [Result.success] 为 [UserEntity]，[Result.failure] 为用户不存在、密码错误或格式错误
     */
    suspend fun login(username: String, password: String): Result<UserEntity> = withContext(Dispatchers.IO) {
        if (!InputValidator.isValidUsername(username))
            return@withContext Result.failure(IllegalArgumentException("用户名格式不正确"))
        if (!InputValidator.isValidPassword(password))
            return@withContext Result.failure(IllegalArgumentException("密码格式不正确"))
        val user = db.userDao().getByUsername(username)
            ?: return@withContext Result.failure(Exception(AppConstants.ERROR_USER_NOT_FOUND))
        val match = PasswordUtil.verify(password, user.password) || user.password == password
        if (!match)
            return@withContext Result.failure(Exception(AppConstants.ERROR_PASSWORD_WRONG))
        if (user.password == password) {
            val updated = user.copy(password = PasswordUtil.encrypt(password))
            db.userDao().update(updated)
        }
        saveCurrentUser(user.id, user.username)
        Logger.d("UserRepository", "Logged in: $username")
        Result.success(user)
    }

    /** 根据 [getCurrentUserId] 从数据库查询当前用户实体，未登录返回 null。 */
    suspend fun getCurrentUser(): UserEntity? = withContext(Dispatchers.IO) {
        val id = getCurrentUserId()
        if (id == -1L) null else db.userDao().getById(id)
    }
}
